File: src/main/kotlin/at/fhhgb/mc/component/sim/sectorizing/QuadTreeSectors.java
package at.fhhgb.mc.component.sim.sectorizing;

import at.fhhgb.mc.component.sim.domain.FloorManager;
import at.fhhgb.mc.component.sim.domain.Matrix;
import at.fhhgb.mc.component.sim.domain.PixelPosition;
import at.fhhgb.mc.component.sim.model.Position;
import at.fhhgb.mc.component.sim.model.recorder.network.AccessPoint;
import at.fhhgb.mc.component.sim.sectorizing.quadtree.AbstractNode;
import at.fhhgb.mc.component.sim.sectorizing.quadtree.AbstractNodeElement;
import at.fhhgb.mc.component.sim.sectorizing.quadtree.point.PointNode;
import at.fhhgb.mc.component.sim.sectorizing.quadtree.point.PointQuadTree;
import at.fhhgb.mc.component.sim.view.DrawingContext;

import java.awt.*;
import java.util.Map;
import java.util.Vector;

/*
* Name: Michael Nigl
*
* ID: 1510455019
*/
public class QuadTreeSectors implements ISectorizing {


    protected PointQuadTree<String> tree;

    protected Vector<AbstractNodeElement<String>> selectedElements = new Vector<AbstractNodeElement<String>>();

    private Position position;

    private boolean matchingSector = false;


    @Override
    public void createSectors(FloorManager floorManager, Graphics g, Matrix m_tMatrix) {
        PointQuadTree<String> tree = new PointQuadTree<String>(new Point(floorManager.getOffsetLeftInPixel(), floorManager.getOffsetTopInPixel()),
                new Dimension(floorManager.getOffsetRightInPixel() - floorManager.getOffsetLeftInPixel(), floorManager.getOffsetBottomInPixel() - floorManager.getOffsetTopInPixel()));
        for (AccessPoint accessPoint : floorManager.getFloor().getAccessPointList()) {
            PixelPosition pixelPosition = floorManager.calculatePixelPositionFromMeter(accessPoint.getPosition().getX(), accessPoint.getPosition().getY());
            tree.insert(pixelPosition.getXPosition(), pixelPosition.getYPosition(), accessPoint.getBSSID());
        }
        PointNode<String> rootNode = tree.getRootNode();
        drawQuadTree(rootNode, g, m_tMatrix);
        matchingSector = false;
    }


    private void drawQuadTree(PointNode<String> node, Graphics g, Matrix matrix) {
        Dimension bounds = node.getBounds();
        Point startCoordinates = node.getStartCoordinates();

        // Draw subnodes
        Map<AbstractNode.Cell, PointNode<String>> subNodes = node.getSubNodes();
        for (PointNode<String> subNode : subNodes.values()) {
            drawQuadTree(subNode, g, matrix);
        }
        // Draw node bounds
        Rectangle rect = new Rectangle(startCoordinates.x, startCoordinates.y, bounds.width, bounds.height);
        if (position != null && !matchingSector && rect.contains(new Point(position.getX(), position.getY()))) {
            DrawingContext.drawCells(rect, g, matrix, true);
            matchingSector = true;
        } else {
            DrawingContext.drawCells(rect, g, matrix, false);
        }
    }

    @Override
    public void addCurrentPosition(Position position) {
        this.position = position;
    }

}

File: src/main/kotlin/at/fhooe/mc/wifipositioning/model/sectoring/QuadTreeSectors.kt
package at.fhooe.mc.wifipositioning.model.sectoring

import at.fhooe.mc.wifipositioning.model.graphics.FloorManager
import at.fhooe.mc.wifipositioning.model.graphics.Matrix
import at.fhooe.mc.wifipositioning.model.building.Position
import at.fhooe.mc.wifipositioning.model.sectoring.quadtree.AbstractNodeElement
import at.fhooe.mc.wifipositioning.model.sectoring.quadtree.point.PointNode
import at.fhooe.mc.wifipositioning.model.sectoring.quadtree.point.PointQuadTree
import at.fhooe.mc.wifipositioning.model.graphics.DrawingContext

import java.awt.*
import java.util.Vector

class QuadTreeSectors : ISectoring {
    protected var tree: PointQuadTree<String>? = null
    protected var selectedElements = Vector<AbstractNodeElement<String>>()
    private var position: Position? = null
    private var matchingSector = false

    override fun createSectors(floorManager: FloorManager, g: Graphics, m_tMatrix: Matrix) {
        val tree = PointQuadTree<String>(Point(floorManager.offsetLeftInPixel, floorManager.offsetTopInPixel),
                Dimension(floorManager.offsetRightInPixel - floorManager.offsetLeftInPixel, floorManager.offsetBottomInPixel - floorManager.offsetTopInPixel))

        for (accessPoint in floorManager.floor!!.accessPoints!!) {
            val pixelPosition = floorManager.calculatePixelPositionFromMeter(accessPoint.position.x, accessPoint.position.y)
            tree.insert(pixelPosition.x, pixelPosition.y, accessPoint.bssid)
        }

        val rootNode = tree.rootNode
        drawQuadTree(rootNode, g, m_tMatrix)
        matchingSector = false
    }


    private fun drawQuadTree(node: PointNode<String>, g: Graphics, matrix: Matrix) {
        val bounds = node.bounds
        val startCoordinates = node.startCoordinates

        // Draw subnodes
        val subNodes = node.subNodes
        for (subNode in subNodes.values) {
            drawQuadTree(subNode, g, matrix)
        }
        // Draw node bounds
        val rect = Rectangle(startCoordinates.x, startCoordinates.y, bounds.width, bounds.height)
        if (position != null && !matchingSector && rect.contains(Point(position!!.x, position!!.y))) {
            DrawingContext.drawCells(rect, g, matrix, true)
            matchingSector = true
        } else {
            DrawingContext.drawCells(rect, g, matrix, false)
        }
    }

    override fun addCurrentPosition(position: Position) {
        this.position = position
    }
}
